### ISSUE TITLE
Add support for IPMI Power Supply sensors

### ISSUE THREAD

**Issue by retrack** [collectd/collectd#735]
The IPMI plugin reports

```
ipmi plugin: sensor_list_add: Ignore sensor PS1 Status0 power_supply (10.1), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
```

The system on which this is running is a Supermicro machine with IPMI, and the sensor list is:

```
# ipmitool sdr
CPU Temp         | 28 degrees C      | ok
PCH Temp         | 43 degrees C      | ok
System Temp      | 24 degrees C      | ok
Peripheral Temp  | 24 degrees C      | ok
FAN1             | 1425 RPM          | ok
FAN2             | 1425 RPM          | ok
FAN3             | 1425 RPM          | ok
FAN4             | 1425 RPM          | ok
Vcore            | 0.69 Volts        | ok
VCCSA            | 0.91 Volts        | ok
VCC PCH          | 1.04 Volts        | ok
VTT              | 1.04 Volts        | ok
VTT MEM          | 0.74 Volts        | ok
3.3V             | 3.26 Volts        | ok
+3.3VSB          | 3.26 Volts        | ok
5V               | 4.99 Volts        | ok
+5VSB            | 5.12 Volts        | ok
12V              | 11.98 Volts       | ok
VBAT             | 3.12 Volts        | ok
V_DIMM           | 1.49 Volts        | ok
PS1 Status       | 0x01              | ok
PS2 Status       | 0x01              | ok
```

Adding support for the Power supply status would permit detection of power supply failure or redundancy status.

**Comment by yadsirhc** [comment]
Any updates on getting this into a future collectd release?  I'm very interested in leveraging this functionality.

**Comment by rrajeevan** [comment]
Hi @rubenk , @octo , any idea when the patch is available? I tried collectd 5.5 but facing the same error.

**Comment by varesa** [comment]
I tried to add support for power supply readings to collectd.
I also noticed, [like on the mailing list](http://mailman.verplant.org/pipermail/collectd/2012-November/005476.html) that just adding a case for IPMI_SENSOR_TYPE_POWER_SUPPLY did not work.

It seems that it stops at OpenIPMI function stand_ipmi_sensor_get_reading():

```
if (sensor->event_reading_type != IPMI_EVENT_READING_TYPE_THRESHOLD)                                                                                                                                                                                            
    /* Not a threshold sensor, it doesn't have readings. */                                                                                                                                                                                                    
    return ENOSYS;
```

In my case `sensor->event_reading_type` is `IPMI_EVENT_READING_TYPE_SENSOR_SPECIFIC`
 
If I use GDB to change the type to THRESHOLD, collectd starts getting values.

ipmitool shows the values correctly:

```
Sensor ID              : Power Supply 1 (0x8)
 Entity ID             : 10.1
 Sensor Type (Discrete): Power Supply
 Sensor Reading        : 200 Watts
 States Asserted       : Power Supply
                         [Presence detected]
```

I'm not sure what to do now

**Comment by shalstea** [comment]
+1

**Comment by octo** [comment]
I've created PR #1936 with my patch. Please leave a comment there if it is working for you. So far, I haven't seen a comment saying that it actually starts reporting anything, so I'm a bit hesitant to merge.

**Comment by rpv-tomsk** [comment]
New features was added to trunk and will be released in collectd-5.8, please check if they are good for you. Feel free to (re)open tickets if needed.

**Comment by johslarsen** [comment]
I have implemented proof-of-concept support for reading discrete IPMI sensors. I agree with @ConiKost's comment in #2014 that the expected behavior would be to record the bit flags, and I have implemented this as a`states` gauge measurement that records the raw bit field values similarly to how `ipmitool sensor` reports these sensors. I have tested this with HP ILO 2,4,5, and it seems to report discrete values nicely for all of these.

Are you interested in exploring such a solution further? Or would such a bit field measurement clash with what collectd (or the IPMI plugin) considers to be a sensor value?

In the [IPMI specification]([https://www.intel.com/content/dam/www/public/us/en/documents/product-briefs/ipmi-second-gen-interface-spec-v2-rev1-1.pdf]) I found a summary of what the bits mean for each type of non-OEM sensors (Table 42-3, p. 505). Instead of a `states` type we could classify the discrete sensors into different types, but I think that reporting anything but bit fields for each of these would in many cases lose a lot of information regardless. For now would think such a generic raw bit field measurement is useful, and then we could save the types for future sensor-specific interpretations of the flags (e.g. a PSU could have somewhat independent state for present/missing, connected/unplugged and working/failed) if collectd ever gets support for something like flag measurements.
